### PR TITLE
Included support for Updatehub library

### PR DESCRIPTION
**Pull request by chtavares592**
This library is a support for the UPDATEHUB platform in zephyr. The library has the ability to query the UPDATEHUB server for the purpose of verifying that it has some update to the board.
Having an update the UPDATEHUB library itself is able to download the update and save it on the card, everything happens correctly the library will restart the card and the new image will start working.

### CODE REVIEW

File: ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h [hunk 1]
```diff
@@ -50,6 +50,7 @@
 #define MBEDTLS_MD_C
 #define MBEDTLS_MD5_C
 #define MBEDTLS_OID_C
+#define MBEDTLS_SSL_MAX_FRAGMENT_LENGTH
```
Comment (otavio) [inline ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h]: This seems unrelated as you made a custom config below. Please remove.

File: ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h [hunk 2]
```diff
@@ -58,6 +59,7 @@
 #define MBEDTLS_SSL_CLI_C
 #define MBEDTLS_SSL_SRV_C
 #define MBEDTLS_SSL_TLS_C
+#define MBEDTLS_PEM_PARSE_C
```
Comment (otavio) [inline ext/lib/crypto/mbedtls/configs/config-mini-tls1_2.h]: Same here.

File: subsys/CMakeLists.txt [hunk 3]
```diff
@@ -13,3 +13,5 @@ add_subdirectory_ifdef(CONFIG_USB                  usb)
 add_subdirectory(random)
 add_subdirectory(storage)
 add_subdirectory_ifdef(CONFIG_SETTINGS             settings)
+add_subdirectory_ifdef(CONFIG_UPDATEHUB updatehub)
+
```
Comment (otavio) [inline subsys/CMakeLists.txt]: Useless newline. Please drop.

File: subsys/Kconfig [hunk 4]
```diff
@@ -32,3 +32,6 @@ source "subsys/random/Kconfig"
 source "subsys/storage/Kconfig"
 
 source "subsys/settings/Kconfig"
+
+source "subsys/updatehub/Kconfig"
+
```
Comment (otavio) [inline subsys/Kconfig]: Useless newline. Please drop.

File: subsys/updatehub/Kconfig [hunk 5]
```diff
@@ -0,0 +1,49 @@
+#
+# Copyright (c) 2018 O.S.Systems
+#
+# SPDX -License-Identifier: Apache-2.0
+#
+
+menu "O.S.Systems Updatehub Library configuration"
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: Please use instead `UpdateHub Firmware Over-the-Air library configuration`

File: subsys/updatehub/Kconfig [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+config UPDATEHUB
+    bool "Updatehub library application settings"
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: UpdateHub Firmware Over-the-Air library support

File: subsys/updatehub/Kconfig [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        selects the APIs needed for the operation of the updatehub library
+
+config UPDATEHUB_POLL_INTERVAL
+    int "Time to poll interval (in minutes)"
+    default 0
+    range 0 43200
+    depends on UPDATEHUB
+    help
+        Set the interval of time that the updatehub library performs the update query.
+        This time interval is zero and 43200 minutes(30 days).
```
Comment (otavio) [inline subsys/updatehub/Kconfig]:
```
Set the interval that the UpdateHub update server will be polled.
```

File: subsys/CMakeLists.txt [hunk 8]
```diff
@@ -13,3 +13,4 @@ add_subdirectory_ifdef(CONFIG_USB                  usb)
 add_subdirectory(random)
 add_subdirectory(storage)
 add_subdirectory_ifdef(CONFIG_SETTINGS             settings)
+add_subdirectory_ifdef(CONFIG_UPDATEHUB updatehub)
```
Comment (otavio) [inline subsys/CMakeLists.txt]: Please follow same indent style from line 15.

File: subsys/updatehub/Kconfig [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    select HTTPS
+    select MBEDTLS_ENABLE_HEAP
+    select NET_APP_SETTINGS
+    select NET_APP
+    select IMG_MANAGER
+    select JSON_LIBRARY
+    default "n"
+    help
+        selects the APIs needed for the operation of the updatehub library
+
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: The options below ought to show only if updatehub support is enabled.

File: subsys/updatehub/device_identity.c [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <soc.h>
+
+#include "device_identity.h"
+
+/*Sample example using serial as unique device-identity.
```
Comment (otavio) [inline subsys/updatehub/device_identity.c]: Fix style.

File: subsys/updatehub/updatehub.c [hunk 11]
```diff
@@ -0,0 +1,888 @@
...
+#include "ssl_crt.h"
+
+#define APP_TIMEOUT K_SECONDS(20)
+#define TCP_RECV_BUFFER_SIZE 16384
+#define URL_SIZE 250
+#define SEND_BUFFER_SIZE 310
+#define HTTP_HEADER_CONNECTION_CLOSE_CRLF "Connection: close\r\n"
+
+// Support https
+#define RESULT_BUF_SIZE MBEDTLS_SSL_MAX_CONTENT_LEN
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Use MBEDTLS_SSL_MAX_CONTENT_LEN directly.

File: subsys/updatehub/updatehub.c [hunk 12]
```diff
@@ -0,0 +1,888 @@
...
+#define HTTPS_STACK_SIZE (2048)
+NET_STACK_DEFINE(HTTPS, https_stack, HTTPS_STACK_SIZE, HTTPS_STACK_SIZE);
+NET_APP_TLS_POOL_DEFINE(ssl_pool, 8);
+
+#define UPDATEHUB_SIGNATURE_SIZE 1024
+#define UPDATEHUB_DECODE_B64_SIZE 512
+#define UPDATEHUB_SHA256_SIZE 32
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Please check those constants.

File: subsys/updatehub/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+zephyr_library_sources_ifdef(CONFIG_UPDATEHUB updatehub.c)
+
+zephyr_link_interface_ifdef(CONFIG_MBEDTLS mbedTLS)
+zephyr_library_link_libraries_ifdef(CONFIG_MBEDTLS mbedTLS)
+
+generate_inc_file_for_target(
+  app 
+  $ENV{UPDATEHUB_UHUPKG_PUBLIC_KEY}
+  ${ZEPHYR_BINARY_DIR}/include/generated/public_key.pem.inc
+)
```
Comment (otavio) [inline subsys/updatehub/CMakeLists.txt]: New line at end.

File: subsys/updatehub/Kconfig [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    select MPU_ALLOW_FLASH_WRITE
+    select NETWORKING
+    select NET_TCP
+    select HTTP
+    select HTTP_CLIENT
+    select NET_APP_NEED_IPV4
+    select NET_DHCPV4
+    select HTTPS
+    select MBEDTLS_ENABLE_HEAP
+    select NET_APP_SETTINGS
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: NET_APP_SETTINGS is not intended to be used. From help:

```
menuconfig NET_APP_SETTINGS
        bool "Set network settings for applications"
        default n
        depends on NET_APP
        help
          Allow IP addresses to be set in config file for
          networking client/server sample applications, or
          some link-layer dedicated settings like the channel.
          Beware this is not meant to be used for proper
          provisioning but quick sampling/testing.
```

File: subsys/updatehub/Kconfig [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    select NET_APP
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: I think we ought to select NET_APP_TLS.

File: subsys/updatehub/Kconfig [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    select IMG_MANAGER
+    select JSON_LIBRARY
+    select SYS_LOG
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: This should be optional, no?

File: subsys/updatehub/ssl_crt.h [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+	"mR7PZ7FWsyNFqUs4gONgvjXz/mS1sHNYGmhLWWUWm2jQXexdzkRpPJ7RTmGXHmz6\r\n"
+	"kQERbOILn6A+iZD3YOmLkIhdb8k7rT0A+/JGteYgDSev5X94OogatqALqHUMIIBX\r\n"
+	"X/wD7XLoYABbBfUPBJ+MHyECAwEAATANBgkqhkiG9w0BAQsFAAOCAQEAY+CBhNag\r\n"
+	"2CN7HztJ4TUQBzC0kWPkT6EJwBj8f1JziHwn6mAqSJJVFBP51NNXAAQpQW1R9Pqv\r\n"
+	"cMjrN63jriFi686U6VXBvIs3eWC/+BWgXKy6RJYUp3x/ISUkS5K/BnarHecaHXXG\r\n"
+	"uwhzQcUDsmOXjQrX2nWGu4HubFf0RA/Ih5YpsyUaeoGxicrGh7A+lHWvtnsYmzBv\r\n"
+	"i90YmN9ak6Qf5bRFwSrHDze+Eq4ow6bXxs4y2dgWWrEyb8GDPP3DUJtTBahDIeAC\r\n"
+	"9Wt8vVinqsnjYOVfdd0gfZO4aat0ojosPMHUGyySnfgNGaEg0hxY5313/hkF02Hq\r\n"
+	"JwnZCTUPqElBtw==\r\n"
+	"-----END CERTIFICATE-----\r\n";
```
Comment (otavio) [inline subsys/updatehub/ssl_crt.h]: Include the certificate from a file.

File: subsys/updatehub/Kconfig (hunk 15 above)
Comment (chtavares592) [reply subsys/updatehub/Kconfig]: NET_APP_TLS is set when the select HTTPS.

```
config HTTPS
    bool
    prompt "HTTPS support" if HTTP && NETWORKING
    default "n" if HTTP && NETWORKING
    select NET_APP_TLS if HTTP && NETWORKING
    depends on HTTP && NETWORKING
    help
      Enables HTTPS support.
```

File: subsys/updatehub/ssl_crt.h (hunk 17 above)
Comment (chtavares592) [reply subsys/updatehub/ssl_crt.h]: Where Can I store?  The same repository of keys?

File: subsys/updatehub/Kconfig (hunk 14 above)
Comment (chtavares592) [reply subsys/updatehub/Kconfig]: But without NET_APP_SETTINGS happens a build error

File: subsys/updatehub/device_identity.c (hunk 10 above)
Comment (chtavares592) [reply subsys/updatehub/device_identity.c]: This style is of uncrustify

File: subsys/updatehub/Kconfig [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+    select HTTPS
+    select MBEDTLS_ENABLE_HEAP
+    select NET_APP_SETTINGS
+    select NET_APP
+    select IMG_MANAGER
+    select JSON_LIBRARY
+    default "n"
+    help
+        The options below ought to show only if updatehub support is enabled.
+
```
[the remark this answers is not in the document]
Comment (chtavares592) [reply subsys/updatehub/Kconfig]: I already change this part

File: subsys/updatehub/CMakeLists.txt (hunk 13 above)
Comment (chtavares592) [reply subsys/updatehub/CMakeLists.txt]: Fixed

File: subsys/updatehub/Kconfig [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+config UPDATEHUB_SYS_LOG 
+    bool ""
+    select SYS_LOG
+    default "y"
+    help
+        TODO
+
+config UPDATEHUB_SYS_LOG_LEVEL
+	int
+	prompt "Updatehub Logging level"
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: UpdateHub

File: subsys/updatehub/certs/ssl_crt.pem [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+mR7PZ7FWsyNFqUs4gONgvjXz/mS1sHNYGmhLWWUWm2jQXexdzkRpPJ7RTmGXHmz6
+kQERbOILn6A+iZD3YOmLkIhdb8k7rT0A+/JGteYgDSev5X94OogatqALqHUMIIBX
+X/wD7XLoYABbBfUPBJ+MHyECAwEAATANBgkqhkiG9w0BAQsFAAOCAQEAY+CBhNag
+2CN7HztJ4TUQBzC0kWPkT6EJwBj8f1JziHwn6mAqSJJVFBP51NNXAAQpQW1R9Pqv
+cMjrN63jriFi686U6VXBvIs3eWC/+BWgXKy6RJYUp3x/ISUkS5K/BnarHecaHXXG
+uwhzQcUDsmOXjQrX2nWGu4HubFf0RA/Ih5YpsyUaeoGxicrGh7A+lHWvtnsYmzBv
+i90YmN9ak6Qf5bRFwSrHDze+Eq4ow6bXxs4y2dgWWrEyb8GDPP3DUJtTBahDIeAC
+9Wt8vVinqsnjYOVfdd0gfZO4aat0ojosPMHUGyySnfgNGaEg0hxY5313/hkF02Hq
+JwnZCTUPqElBtw==
+-----END CERTIFICATE-----
```
Comment (otavio) [inline subsys/updatehub/certs/ssl_crt.pem]: Newline at end.

Comment (chtavares592) [reply subsys/updatehub/certs/ssl_crt.pem]: fixed

File: subsys/updatehub/device_identity.c [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+/*
+ * @brief Sample example using serial as unique device-identity.
+ *
+ * @details This is an implementation example for creating a unique identifier for this board.
+ * Depending on the board or the user's choice, another implementation is required, but the board ID must be unique
+ * and the functions names must be keep.
+ * DEVICE_ID_BASE: beginning of HW UID registers
+ * DEVICE_ID_LENGTH: length of HW UID registers in 32-bit words
+ *
+ */
```
Comment (otavio) [inline subsys/updatehub/device_identity.c]: Here this should be a normal comment as this is not indexed by the documentation.

I think it might be as:

```
/*
 * This is SoC / board specific. The device ID must be unique.
 */
```

File: subsys/updatehub/device_identity.c [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+#if defined(CONFIG_SOC_SERIES_NRF52X)
```
Comment (otavio) [inline subsys/updatehub/device_identity.c]: A new comment here might be:

```
/*
 * DEVICE_ID_BASE: points to the base address of hardware ID registers
 * DEVICE_ID_LENGTH: number or 32-bit words to read
 */

File: subsys/updatehub/device_identity.h [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+#ifndef _DEVICE_IDENTITY_H__
+#define _DEVICE_IDENTITY_H__
+
+/*
+   @brief Gets the identity of device
+ 
+   @details Gets the device identity sets by device identity init function
+ */
+
```
Comment (otavio) [inline subsys/updatehub/device_identity.h]: Drop this extra line.

File: subsys/updatehub/device_identity.h [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * @brief Initializes the unique identity of the device
+ *
+ * @details When the UpdateHub starts this function sets the unique device identifier
+ * that will be used know that device is updating.
+ *
+ * @param struct device ctx
+ */
+
```
Comment (otavio) [inline subsys/updatehub/device_identity.h]: Same here.

File: subsys/updatehub/device_identity.h [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+char *device_identity_get(void);
+
+/**
+ * @brief Initializes the unique identity of the device
+ *
+ * @details When the UpdateHub starts this function sets the unique device identifier
+ * that will be used know that device is updating.
+ *
+ * @param struct device ctx
```
Comment (otavio) [inline subsys/updatehub/device_identity.h]: ctx?

File: subsys/updatehub/updatehub.h [hunk 26]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2018 O.S.Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file 
+ *
+ * @brief UpdateHub implementation for Zephyr.
```
Comment (otavio) [inline subsys/updatehub/updatehub.h]: Please make it closer to the content you used inside the help.

File: subsys/updatehub/updatehub.h [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+#ifndef _UPDATEHUB_H_
+#define _UPDATEHUB_H_
+
+#define UPDATEHUB_SERVER "10.5.3.14"
```
Comment (otavio) [inline subsys/updatehub/updatehub.h]: This must be configured inside kconfig.

File: subsys/updatehub/updatehub.h [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+#define UPDATEHUB_PORT 8080
```
Comment (otavio) [inline subsys/updatehub/updatehub.h]: This can die. You can move your development server to the normal port so you don't need to override it.

File: subsys/updatehub/updatehub.c [hunk 29]
```diff
@@ -0,0 +1,851 @@
...
+#include <net/net_app.h>
+#include <net/net_event.h>
+#include <net/net_mgmt.h>
+#include <net/http.h>
+#include <json.h>
+#include <flash.h>
+#include <dfu/mcuboot.h>
+#include <dfu/flash_img.h>
+
+#include "updatehub.h"
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: It should import from public (<...>)

File: subsys/updatehub/updatehub.c [hunk 30]
```diff
@@ -0,0 +1,851 @@
...
+#include "updatehub_json.h"
+#include "device_identity.h"
+
+#define APP_TIMEOUT K_SECONDS(20)
+#define TCP_RECV_BUFFER_SIZE 16384
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Use MBEDTLS_SSL_MAX_CONTENT_LEN directly.

File: subsys/updatehub/updatehub.c [hunk 31]
```diff
@@ -0,0 +1,851 @@
...
+#include <json.h>
+#include <flash.h>
+#include <dfu/mcuboot.h>
+#include <dfu/flash_img.h>
+
+#include "updatehub.h"
+#include "updatehub_json.h"
+#include "device_identity.h"
+
+#define APP_TIMEOUT K_SECONDS(20)
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]:
```
subsys/net/lib/http/Kconfig:config HTTP_CLIENT_NETWORK_TIMEOUT
subsys/net/lib/http/http_client.c:#define HTTP_NETWORK_TIMEOUT 
K_SECONDS(CONFIG_HTTP_CLIENT_NETWORK_TIMEOUT)
```

So do the same thing. Add an entry inside kconfig and use HTTP_CLIENT_NETWORK_TIMEOUT as default.

File: subsys/updatehub/updatehub.c [hunk 32]
```diff
@@ -0,0 +1,851 @@
...
+#define TCP_RECV_BUFFER_SIZE 16384
+#define URL_SIZE 250
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Use `MAX_URL_SIZE`

File: subsys/updatehub/updatehub.c [hunk 33]
```diff
@@ -0,0 +1,851 @@
...
+					 char *package_uid)
+{
+	snprintk(ctx->http_client.url, URL_SIZE,
+		 "https://%s:%d/%s/%s/packages/%s/objects/%s", CONFIG_UPDATEHUB_SERVER,
+		 CONFIG_UPDATEHUB_PORT, updatehub_url(UPDATEHUB_DOWNLOAD),
+		 response->product_uid, package_uid,
+		 response->objects[0].sha256sum);
+
+	/* Stores url download at download_http */
+	memcpy(download_http, ctx->http_client.url, strlen(ctx->http_client.url) + 1);
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: The `memcpy` seems useless as `snprintk` is writing it.

File: include/dfu/updatehub.h [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+	UPDATEHUB_NO_UPDATES,  /** Returns when the board no have updates to do*/
+    UPDATEHUB_NONE_RESPONSE, /** None message*/
+	UPDATEHUB_NETWORKING_ERROR, /** Returns when occurs any fails on the networking service */
+	UPDATEHUB_SIGNATURE_ERROR, /** Returns when occurs any fails at to check the package signature */
+	UPDATEHUB_SUPPORTED_HARDWARE_ERROR, /** Returns when the hardware is not suportted for the package downloaded */
+	UPDATEHUB_METADATA_ERROR, /** Returns when the updatehub api cannot parse the metadata */
+	UPDATEHUB_DOWNLOAD_ERROR, /** Returns when occurs any fails at to download the package or image */
+	UPDATEHUB_INSTALL_ERROR, /** Returns when the image cannot be written on the flash */
+	UPDATEHUB_FLASH_MEMORY_ERROR, /** Returns when occurs some error of flash initialization */
+};
```
Comment (otavio) [inline include/dfu/updatehub.h]: The indenting is wrong. Also, drop the returns when in every entry.

File: include/dfu/updatehub.h [hunk 35]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * @brief Responses messages from UpdateHub library.
+ *
+ * @details These messages are used to inform the server and the user about the process status of 
+ * the UpdateHub library and also used to standardize the errors that may occur.
+ *
+ */
+enum updatehub_response {
+	UPDATEHUB_SUCCESS = 0, /** When everithing runs good */
```
Comment (otavio) [inline include/dfu/updatehub.h]: `UPDATEHUB_SUCCESS` is going to be used inside the run?

If this is the case, consider using `UPDATEHUB_UPDATE_COMPLETE` might be clearer.

File: subsys/updatehub/Kconfig [hunk 36]
```diff
@@ -0,0 +1,83 @@
...
+if UPDATEHUB
+config UPDATEHUB_SERVER
+    string "Set the address server of UpdateHub"
+    default "api.updatehub.io"
+    depends on UPDATEHUB
+
+config UPDATEHUB_PORT
+    int "Set the port of UpdateHub"
+    default "80"
+    depends on UPDATEHUB
```
Comment (otavio) [inline subsys/updatehub/Kconfig]: SERVER and PORT can die.

File: include/dfu/updatehub.h [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+ * @brief Responses messages from UpdateHub library.
+ *
+ * @details These messages are used to inform the server and the user about the process status of
+ * the UpdateHub library and also used to standardize the errors that may occur.
+ *
+ */
+enum updatehub_response {
+	UPDATEHUB_HAS_UPDATE = 0,               /** When has update to do*/
+	UPDATEHUB_NO_UPDATES,                   /** When the board no have updates to do*/
+	UPDATEHUB_NONE_RESPONSE,                /** None message */
```
Comment (otavio) [inline include/dfu/updatehub.h]: The NONE state should die.

File: include/dfu/updatehub.h [hunk 38]
```diff
@@ -0,0 +1,70 @@
...
+	UPDATEHUB_NETWORKING_ERROR,             /** When occurs any fails on the networking service */
+	UPDATEHUB_SIGNATURE_ERROR,              /** When occurs any fails at to check the package signature */
+	UPDATEHUB_SUPPORTED_HARDWARE_ERROR,     /** When the hardware is not suportted for the package downloaded */
+	UPDATEHUB_METADATA_ERROR,               /** When the updatehub api cannot parse the metadata */
+	UPDATEHUB_DOWNLOAD_ERROR,               /** When occurs any fails at to download the package or image */
+	UPDATEHUB_INSTALL_ERROR,                /** When the image cannot be written on the flash */
+	UPDATEHUB_FLASH_MEMORY_ERROR,           /** When occurs some error of flash initialization */
+};
```
Comment (otavio) [inline include/dfu/updatehub.h]:
```
enum updatehub_response {
        UPDATEHUB_HAS_UPDATE = 0,               /** Has an update available */
        UPDATEHUB_NO_UPDATE,                    /** No update available */
        UPDATEHUB_NETWORKING_ERROR,             /** Fail to connect to the UpdateHub server */
        UPDATEHUB_SIGNATURE_ERROR,              /** Invalid update package signature */
        UPDATEHUB_INCOMPATIBLE_HARDWARE,        /** Incompatible hardware */
        UPDATEHUB_METADATA_ERROR,               /** Fail to parse the metadata */
        UPDATEHUB_DOWNLOAD_ERROR,               /** Fail while downloading the update package */
        UPDATEHUB_INSTALL_ERROR,                /** Fail while installing the update package */
        UPDATEHUB_FLASH_INIT_ERROR,             /** Fail to initilialize the flash */
};
```

File: subsys/updatehub/updatehub.c [hunk 39]
```diff
@@ -0,0 +1,976 @@
...
+	u8_t send_buffer[SEND_BUFFER_SIZE];
+	size_t http_content_size;
+	size_t downloaded_size;
+	int download_status;
+};
+
+struct updatehub_context {
+	struct flash_img_context flash_img_ctx;
+	struct updatehub_http_client http_client;
+	char *error_msg;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Not needed. Instead, store the error code of the enum.

File: subsys/updatehub/updatehub.c [hunk 40]
```diff
@@ -0,0 +1,976 @@
...
+		return "products";
+	case UPDATEHUB_REPORT:
+		return "report";
+	default:
+		SYS_LOG_ERR("Invalid url");
+		return NULL;
+	}
+}
+
+static const char *status_execution(enum updatehub_status status)
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: This is not status but state.

File: subsys/updatehub/updatehub.c [hunk 41]
```diff
@@ -0,0 +1,976 @@
...
+	int i;
+
+	for (i = 0; i < metadata_sh_array->supported_hardware_len; i++) {
+		if (strcmp(metadata_sh_array->supported_hardware[i], CONFIG_BOARD) == 0) {
+			return 0;
+		}
+	}
+	return -UPDATEHUB_SUPPORTED_HARDWARE_ERROR;
+}
+
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]:
```
static int is_compatible(struct updatehub_probe_recv_metadata_sh_array *metadata_sh_array)
{
        int i;
        int ret = -1;

        for (i = 0; i < metadata_sh_array->supported_hardware_len; i++) {
                if (strcmp(metadata_sh_array->supported_hardware[i], CONFIG_BOARD) == 0) {
                        ret = 0;
                        break;
                }
        }

        return ret;
}
```

File: subsys/updatehub/updatehub.c [hunk 42]
```diff
@@ -0,0 +1,976 @@
...
+			end--;
+			break;
+		}
+	}
+
+	memcpy(value, &headers[start], end - start);
+	value[end - start] = '\0';
+}
+
+static int package_signature_verify(struct updatehub_context *ctx, unsigned char *metadata_hash)
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: check_package_signature(...)

File: subsys/updatehub/updatehub.c [hunk 43]
```diff
@@ -0,0 +1,976 @@
...
+	char buffer[2];
+	u8_t *body_data = NULL;
+	u8_t image_hash[UPDATEHUB_SHA256_SIZE];
+	size_t body_len = 0;
+
+	struct updatehub_context *ctx = user;
+
+	if (ctx->http_client.http_content_size == 0) {
+		if (http->http.rsp.body_found == 0) {
+			ctx->error_msg = updatehub_response(UPDATEHUB_DOWNLOAD_ERROR);
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: ctx->response

File: subsys/updatehub/updatehub.c [hunk 44]
```diff
@@ -0,0 +1,976 @@
...
+		k_sem_give(ctx->sem);
+	}
+
+	return;
+
+error:
+	k_sem_give(ctx->sem);
+}
+
+static int install_update(struct updatehub_context *ctx)
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]:
```
static updatehub_response _install_update(...)

File: subsys/updatehub/updatehub.c [hunk 45]
```diff
@@ -0,0 +1,960 @@
...
+}
+
+static void build_http_request(struct updatehub_context *ctx)
+{
+	/* Builds http_request Probe and Report */
+	memset(&ctx->http_client.req, 0, sizeof(ctx->http_client.req));
+	ctx->http_client.req.method = HTTP_POST;
+	ctx->http_client.req.url =  ctx->http_client.url;
+	ctx->http_client.req.protocol = " " HTTP_PROTOCOL;
+	ctx->http_client.req.header_fields = HTTP_HEADER_CONNECTION;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Add the header here.

File: subsys/updatehub/updatehub.c [hunk 46]
```diff
@@ -0,0 +1,960 @@
...
+		k_sem_give(ctx->sem);
+	}
+
+	return;
+
+error:
+	k_sem_give(ctx->sem);
+}
+
+static int updatehub_response_install_update(struct updatehub_context *ctx)
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: static updatehub_response     install_update(...)

File: subsys/updatehub/updatehub.c [hunk 47]
```diff
@@ -0,0 +1,960 @@
...
+		 UPDATEHUB_PORT, url(UPDATEHUB_DOWNLOAD),
+		 CONFIG_UPDATEHUB_PRODUCT_UID, probe_config.package_uid,
+		 probe_config.sha256sum_image);
+
+	/* Builds the http_request download*/
+	memset(&ctx->http_client.req, 0, sizeof(ctx->http_client.req));
+	ctx->http_client.req.method = HTTP_GET;
+	ctx->http_client.req.url = ctx->http_client.url;
+	ctx->http_client.req.protocol = " " HTTP_PROTOCOL;
+	ctx->http_client.req.header_fields = HTTP_HEADER_CONNECTION;
```
Comment (otavio) [inline subsys/updatehub/updatehub.c]: Why not reuse the method to create the request?

File: subsys/updatehub/updatehub_internal.h [hunk 48]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+enum updatehub_status {
+	UPDATEHUB_STATUS_EXEC_DOWLOADING = 0,
+	UPDATEHUB_STATUS_EXEC_DOWNLOADED,
+	UPDATEHUB_STATUS_EXEC_INSTALLING,
+	UPDATEHUB_STATUS_EXEC_INSTALLED,
+	UPDATEHUB_STATUS_EXEC_REBOOTING,
+	UPDATEHUB_STATUS_EXEC_ERROR,
+};
```
Comment (otavio) [inline subsys/updatehub/updatehub_internal.h]: state

File: include/dfu/updatehub.h [hunk 49]
```diff
@@ -0,0 +1,70 @@
...
+	UPDATEHUB_HAS_UPDATE = 0,               /** Has an update available */
+	UPDATEHUB_NO_UPDATE,                    /** No update available */
+	UPDATEHUB_NETWORKING_ERROR,             /** Fail to connect to the UpdateHub server */
+	UPDATEHUB_SIGNATURE_ERROR,              /** Invalid update package signature */
+	UPDATEHUB_INCOMPATIBLE_HARDWARE,        /** Incompatible hardware */
+	UPDATEHUB_METADATA_ERROR,               /** Fail to parse the metadata */
+	UPDATEHUB_DOWNLOAD_ERROR,               /** Fail while downloading the update package */
+	UPDATEHUB_INSTALL_ERROR,                /** Fail while installing the update package */
+	UPDATEHUB_FLASH_INIT_ERROR,             /** Fail to initilialize the flash */
+	UPDATEHUB_OK,
```
Comment (otavio) [inline include/dfu/updatehub.h]:
```
... /** It was Ok */

File: include/dfu/updatehub.h [hunk 50]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * @brief Send a HTTP request to UpdateHub server for verify if has updates.
+ *
+ * @details The updatehub_probe function has the responsibility to connect on UpdateHub server
+ * and asks if there is some update to do.
+ *
+ * @return Return 0 if has update, 1 if the board already updated, and < 0 if error.
+ */
+int updatehub_probe();
```
Comment (otavio) [inline include/dfu/updatehub.h]: enum.

File: include/dfu/updatehub.h [hunk 51]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * @brief Initializes the installation of the image in the memory of the board
+ *
+ * @details After called updatehub_probe function and returns success can be call updatehub_update function
+ * has the responsibility to initialize the flash where the update will be installed, download the image and
+ * install at board.
+ *
+ * @return Return 0 if success, and < 0 if error.
+ */
+int updatehub_update();
```
Comment (otavio) [inline include/dfu/updatehub.h]: enum.

File: subsys/updatehub/updatehub_internal.h [hunk 52]
```diff
@@ -0,0 +1,79 @@
...
+#ifndef __UPDATEHUB_JSON_H__
+#define __UPDATEHUB_JSON_H__
+
+enum updatehub_url {
+	UPDATEHUB_PROBE = 0,
+	UPDATEHUB_DOWNLOAD,
+	UPDATEHUB_REPORT,
+};
+
+enum updatehub_status {
```
Comment (otavio) [inline subsys/updatehub/updatehub_internal.h]: state

**Comment by otavio** [comment]
Too much noise; will ask for a new PR so we start with clean history.
